### PR TITLE
sys: added Arduino API to RIOT

### PR DESCRIPTION
**Pull request by haukepetersen**
I know we are in feature freeze currently, but I just had an idea tonight and wanted to try it out - seems like it worked...

With this PR, RIOT is able to run un-touched Arduino sketches! All you have to do is create a RIOT application, put in a RIOT Makefile (with `USEMODULE += arduino`), copy some sketches into this folder and say `make all flash term` -> done.

At the moment, the sketches have to have the file ending `*.sketch` to be found by the RIOT buildsystem, but this could be changed easily...

Also this PR does (of course?) not contain the complete Arduino API. Implement so far are:
- `pinMode`, `digitalRead`, `digitalWrite`
- `delay`
- `Serial` (partly)

But with this as base, implementing most parts of the Arduino API is trivial...

### CODE REVIEW

**Comment by haukepetersen** [comment]
Forgot to state: Currently there is only a configuration for the `stm32f4discovery`. The configuration contains only the pin mapping, so it can easily created for other boards later

**Comment by authmillenon** [comment]
Weren't you the one that always argued against the Arduino API ;)?

**Comment by kaspar030** [comment]
Nice one!

**Comment by authmillenon** [comment]
Some READMEs for the examples would be nice.

**Comment by emmanuelsearch** [comment]
awesome!

**Comment by haukepetersen** [comment]
> Weren't you the one that always argued against the Arduino API ;)?

I still am! IMHO the API is (i) inconsistent, (ii) at time confusing, (iii) in-efficient, (iv) and ugly (though this is of course a very subjective perception :-) ).

I created this for basically two reasons:
1. to make it easy for people to move from Arduino to RIOT -> follow the learning curve...
2. show the power of our peripheral drivers -> putting the Arduino API on top is a no-brainer...

**Comment by haukepetersen** [comment]
Oh, and READMEs will follow - but I am not really planning to put work in this again before the release is out :-)

File: sys/arduino/serialport.cpp [hunk 1]
```diff
@@ -0,0 +1,215 @@
...
+
+size_t SerialPort::print(float val)
+{
+    return print(val, 2);
+}
+
+size_t SerialPort::print(float val, int format)
+{
+    /* TODO: make this pathetic code more efficient and good looking... */
+    char buf[64];
```
Comment (Kijewski) [inline sys/arduino/serialport.cpp]:
```c
printf("%.*f", format, val);
```

Comment (haukepetersen) [inline sys/arduino/serialport.cpp]: ha, interesting! I new there must be something better :-)

**Comment by temmihoo** [comment]
So cool! I'll keep this in mind so that I can teach new programmers a path from Arduino to better world!

**Comment by haukepetersen** [comment]
rebased and optimized `SerialPort::print(float val, int format)`

File: dist/tools/arduino/pre_build.sh [hunk 2]
```diff
@@ -0,0 +1,24 @@
+#!/bin/sh
+
+# check if at least the application dir and one sketch is given
+if [ $# -lt 2 ]
+then
+    echo "[Arduino pre-build] Error: not enough arguments given"
+    exit 1
+fi
+
+# 'comsume' the application and arduino directories (first argument)
```
Comment (Kijewski) [inline dist/tools/arduino/pre_build.sh]: "co**n**sume"?

Comment (haukepetersen) [inline dist/tools/arduino/pre_build.sh]: yap :-)

**Comment by emmanuelsearch** [comment]
So how about some short initial README? What else is mandatory before merging this? It's a new functionality, so it could be extended/improved later, I guess?

**Comment by haukepetersen** [comment]
fixed some issues, added a significant amount of documentation (readme for example, doxygen) and cut down the example code to a single example.

File: examples/arduino_hello-world/hello-world.sketch [hunk 3]
```diff
@@ -0,0 +1,62 @@
+/*
+  Arduino Hello-World @ RIOT
+  Prints 'Hello Arduino!' once on the serial port during startup, toggles the
+  default LED twice each seconds and echoes incoming characters on the serial
+  port.
+ */
+
+// Per convention, RIOT defines a macro mapping the an on-board LED if
```
Comment (cgundogan) [inline examples/arduino_hello-world/hello-world.sketch]: ..`the an`.. should it be `for the` ?

**Comment by cgundogan** [comment]
flashed this branch on a `stm32f4discovery` board and the arduino example was echoing my input as intended. Code-wise everything looks okay. So from my side this is an ACK.

File: examples/arduino_hello-world/hello-world.sketch (hunk 3 above)
Comment (haukepetersen) [inline examples/arduino_hello-world/hello-world.sketch]: not quite, its a missing comma after `macro` :-)

**Comment by haukepetersen** [comment]
fixed @cgundogan's comment and squashed it right in.

File: sys/arduino/doc.txt [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+ *
+ * This module enables users to run unmodified Arduino sketches in RIOT. For
+ * this we aim at supporting the full Arduino API.
+ *
+ * The support of the Arduino API in RIOT is useful for multiple reasons:
+ * - starting point for beginners
+ * - makes it easy to move from Arduino to RIOT
+ * - use Ardunio device drivers in RIOT
+ * - is fun to implement :-)
+ *
```
Comment (emmanuelsearch) [inline sys/arduino/doc.txt]: add: make it easy/possible to run existing Arduino sketches on non-Arduino hardware?

File: sys/arduino/doc.txt [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+ * BOARD ?= YOUR_TARGET_PLATFORM
+ * RIOTBASE ?= PATH_TO_RIOT_ROOT
+ *
+ * USEMODULE += arduino
+ *
+ * include $(RIOTBASE)/Makefile.include
+ * @endcode
+ *
+ * -# copy you Arduino sktech(es) into your application folder. Currently they
+ *    must have the file ending `*.sketch` to be processed. *
```
Comment (emmanuelsearch) [inline sys/arduino/doc.txt]: remove * at the end of the sentence.

**Comment by emmanuelsearch** [comment]
I happily double @cgundogan's ACK ;)

File: sys/arduino/doc.txt (hunk 4 above)
Comment (haukepetersen) [inline sys/arduino/doc.txt]: This was implicitly part of 'makes it easy to move from...', but I'll include it separately..

**Comment by haukepetersen** [comment]
address comments by @emmanuelsearch, waiting for Travis one more time.

**Comment by voxadam** [comment]
Great work @haukepetersen! 

I agree that the Arduino API leaves much to be desired I agree that providing a layer like this in RIOT drastically lowers the bar for entry. @temmihoo is right, it's a great way to provide people with a bit of Arduino knowledge a path to a better, more capable place.

**Comment by temmihoo** [comment]
> 
> On 09 Dec 2015, at 05:36, Adam Hunt <notifications@github.com> wrote:
> 
> Great work @haukepetersen!
> 


Indeed, thx Hauke! 

Lets not leave this at this, I mean there are more apis in Arduino core...
> I agree that the Arduino API leaves much to be desired I agree that providing a layer like this in RIOT drastically lowers the bar for entry. @temmihoo is right, it's a great way to provide people with a bit of Arduino knowledge a path to a better, more capable place.
> 

This compatibility layer fulfills an old ELL-i dream of providing a path set with stepping stones for beginners to start from Arduino and progress through concepts changing not too many things at once.

I see it something like this:

Arduino with its own IDE
Arduino with $EDITOR
... add git
... add Github personal repository
... add collaboration workflows
RIOT OS with Arduino runs most of previous work
... add basic networking through RIOT API
... add testing
... add refactoring
... refactor Arduinoness out of the way
... add thinking of constraints
... ... ...
PROFIT

I see the DVCS skills as most important and would like to introduce them to budding programmers as soon as possible. I have found out that just practicing with meaningless text files does not work so I have waited until we have some context (here: the led blinkers).

Again: thanks Hauke.

 - t

**Comment by haukepetersen** [comment]
Thanks guys! Of course there is much to do/add in terms of API completeness, but this should be easy to add step-by-step. I am about to open PRs for analog read/write, stay tuned :-)

**Pull request merged**